## ALS Spinal Cord differential expression and molecular QTLs

Jack Humphrey 2019-2021

All Rmarkdown workbooks written for the project.

Most workbooks are presented as raw .Rmd (in `scripts/` and knitted HTML files (in `html/`).

## Data QC and setup

[RNA-seq QC](https://jackhump.github.io/ALS_SpinalCord_QTLs/html/ALS_covariates_setup.html)

WGS QC pipeline hosted [here](https://github.com/jackhump/WGS-QC-Pipeline)

[Building sets of marker genes](https://github.com/jackhump/ALS_SpinalCord_QTLs/blob/master/scripts/ALS_marker_lists.Rmd)
 
## Differential expression

[Running differential expression](https://jackhump.github.io/ALS_SpinalCord_QTLs/html/ALS_differential_expression.html)

[Volcano plots, GSEA, marker gene overlaps](https://jackhump.github.io/ALS_SpinalCord_QTLs/html/ALS_deg_functions.html)

[Cell-type deconvolution](https://jackhump.github.io/ALS_SpinalCord_QTLs/html/ALS_deconvolution.html)

## Co-expression networks

[Building co-expression networks, associations with traits and GO](https://jackhump.github.io/ALS_SpinalCord_QTLs/html/ALS_WGCNA.html)

## Quantitative Trait Loci (QTLs)

[QTL results per region and sharing with GTEx](https://jackhump.github.io/ALS_SpinalCord_QTLs/html/QTL_results.Rmd)

[Colocalisation with ALS GWAS](https://jackhump.github.io/ALS_SpinalCord_QTLs/html/ALS_COLOC.Rmd)

[Transcriptome-wide association study - TWAS](https://jackhump.github.io/ALS_SpinalCord_QTLs/html/ALS_TWAS.Rmd)

## Gene prioritisation

[C9orf72 and ATXN3](https://jackhump.github.io/ALS_SpinalCord_QTLs/html/ATXN3_C9orf72.Rmd)

[Transcript plots](https://jackhump.github.io/ALS_SpinalCord_QTLs/html/ALS_locus_plots.Rmd)

[Localising genes to cell-types](https://jackhump.github.io/ALS_SpinalCord_QTLs/html/ALS_genes_celltypes.Rmd)
