---
title: "ALS deconvolution"
author: "Jack Humphrey"
date: "`r Sys.Date()`"
output: 
 rmarkdown::html_document:
   theme: spacelab
   highlight: haddock
   code_folding: hide
   toc: true
   toc_float: true
   smooth_scroll: true
   number_sections: false
   self_contained: true  
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(patchwork)
library(UpSetR)

# remotes::install_github("oganm/markerGeneProfile")
# remotes::install_github('youngjaewoo/DDSClassifier')
# https://github.com/ellispatrick/CortexCellDeconv
library(dtangle)
library(MuSiC)
#library(DSA)
#library(tidyr)
#library(dplyr)
#library(ggplot2)
library(edgeR)
library(limma)
library(grid)
library(MASS)
library(tidyverse)
library(ggrastr)
library(statmod)
#library(markerGeneProfile)
#library(declassification)
#library(glmnet)
# library(amritr)
library(RColorBrewer)
# library(SuppDists)
# library(plotrix)

#library(corpcor)

library(reshape2)

#library(ggsci)

library(ggpubr)

theme_jh <- function () { 
    theme_bw(base_size=10) %+replace% 
        theme(
          panel.grid = element_blank(),
          strip.background = element_blank(),
          #panel.border = element_blank(),
          axis.line = element_line(),
          #text = element_text(color = "black"), 
          strip.text = element_text(color = "black"),
          axis.text = element_text(colour = "black"),
            panel.background  = element_blank(),
            plot.background = element_rect(fill="white", colour=NA), 
            legend.background = element_rect(fill="transparent", colour=NA),
            legend.key = element_rect(fill="transparent", colour=NA)
        )
}

rerun <- FALSE

```

```{r folders, echo=TRUE}
work_dir = here::here("data/collate_samples/")

plots_folder = here::here("docs/ALS_deconvolution/plots/")
if(!dir.exists(plots_folder)){dir.create(plots_folder)}
#deconv_github = "/Users/jack/software/CortexCellDeconv/"
deconv_github <- "/Users/Jack/google_drive/Work/Mount_Sinai/Misc/CortexCellDeconv/"
```


## Input our expression data

Don't apply any normalising or scaling


```{r expression, echo=TRUE}

deconvolution_input_file <- here::here("data/deconvolution/input_expression.RData")
if(rerun){
load(here::here("data/feb_2020/gene_counts_no_tags.RData") )
load(here::here("data/feb_2020/gene_tpm_no_tags.RData") )

metadata <- read_tsv( here::here("data/differential_expression/ALS/ALS_all_differential_expression_support.tsv") )

# filter out non-CNS samples
tissues <- c("Lumbar_Spinal_Cord", "Cervical_Spinal_Cord", "Thoracic_Spinal_Cord")

# remove low RIN samples - for good
metadata <- filter(metadata, tissue %in% tissues)

#Gene name for the voom table 
gencode_30 <- read_tsv(here::here("data/misc/gencode.v30.tx2gene.tsv.gz")) %>%
  janitor::clean_names() %>%
  dplyr::select(ensembl = geneid, symbol = genename) %>%
  dplyr::distinct() %>%
  dplyr::mutate(ensembl =  str_split_fixed(ensembl, "\\.", 2)[,1] ) %>%
  dplyr::distinct() 

genes_loc <- as.data.frame(genes_counts[, metadata$sample])
#rm(genes_counts)

# remove lowly expressed genes
keep.exp <- rowSums(cpm(genes_loc) > 1) >= ceiling(0.5*ncol(genes_loc))



genes_loc <- genes_loc[keep.exp,]
print(nrow(genes_loc))
# QN is bad!
#genes_loc <-  EDASeq::betweenLaneNormalization(as.matrix(genes_loc),which = "full")

# normalisation is ok..
#norm_loc <- calcNormFactors(genes_loc, method = "TMM") #normalized counts with TMM
#dge <- DGEList(counts=genes_loc, samples=data.frame(row.names = colnames(genes_loc), sample = colnames(genes_loc)), norm.factors = norm_loc) #design_loc <- model.matrix(as.formula(design_formula), data = support_loc)
#genes_counts_voom_loc <- voom(dge)

#genes_counts_voom[1:5,1:8]
geneExpr_our = as.data.frame(genes_loc)
geneExpr_our$ensembl = row.names(geneExpr_our)
geneExpr_our <- merge(geneExpr_our, gencode_30, by = "ensembl")
#Remove duplicates 
geneExpr_our <- geneExpr_our[! duplicated(geneExpr_our$symbol),]
dim(geneExpr_our) # 19730   276
rownames(geneExpr_our) = geneExpr_our$symbol
geneExpr_our$ensembl = NULL
geneExpr_our$symbol = NULL

#head(geneExpr_our)

save(geneExpr_our, file = deconvolution_input_file)
}else{
load(deconvolution_input_file)
}
# row scaling is bad!
#geneResid_our = geneExpr_our - rowMeans(geneExpr_our)
```


## Load Darmanis data
Ben Barres and Stephen Quake paper. 

Reference dataset from Darmanis et al. "A survey of human brain transcriptome diversity at the single cell level." 
https://www.pnas.org/content/112/23/7285.long#sec-7

MuSic requires replicate donors for each  cell type to measure variability between donors.
Darmanis luckily took single cells from multiple brains.

DarmanisCells from GSE67835 single cell data.

```{r DarmanisCells, echo=FALSE,warning=FALSE,messsage=FALSE}

# Patrick has already collated the counts and removed any hybrid, fetal or OPC cells
#deconv_github <- "~/google_drive/Work/Mount_Sinai/CortexCellDeconv/"
deconv_github <- "/Users/Jack/google_drive/Work/Mount_Sinai/Misc/CortexCellDeconv/"
load(paste0(deconv_github, "CellTypeDeconvAnalysis/Data/Darmanis_counts.RData"))

# prepare metadata
# downloaded from SRA: https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA281204&o=acc_s%3Aa 
# click Download metdata
darmanis_metadata_file <- here::here("data/Darmanis/Darmanis_GSE67835_metadata.txt")

# extract fields from metadata
darmanis_metadata <- 
  readr::read_csv(darmanis_metadata_file) %>% 
  janitor::clean_names() %>%
  dplyr::select(cellID = sample_name, cellType = cell_type, sampleID = experiment_sample_name )

darmanis_metadata <- filter(darmanis_metadata, cellID %in% colnames(Darmanis))

DarmanisCells <- darmanis_metadata$cellType
names(DarmanisCells) <- darmanis_metadata$cellID

```


Try to factorise code into functions

```{r createMarkers, echo=FALSE,warning=FALSE,messsage=FALSE}
# take reference dataset (Darmanis) and create markers
# differential expression - compare each cell type to the mean
compareReferenceCells <- function(bulk_expression, reference_expression, reference_meta){
  # only use reference genes present in bulk
  int <- intersect(rownames(bulk_expression),rownames(reference_expression))
  y <- DGEList(counts= reference_expression[int,])
  y <- calcNormFactors(y,method = 'TMM', Acutoff =  quantile(log2(reference_expression[,1]/sum(reference_expression[,1])),0.75))
  
  reference_expression <- cpm(y, log=FALSE)
  reference_expressionMean = sapply(unique(reference_meta),function(x)rowMeans(reference_expression[,names(which(reference_meta==x))]))
  int = intersect(rownames(bulk_expression),rownames(reference_expressionMean))
  design = model.matrix(~.-1,data.frame(cell= reference_meta[colnames(reference_expression)]))
  colnames(design) = gsub('cell','',colnames(design))
  v <- voom(reference_expression[int,rownames(design)], design, plot=FALSE)
  fit <- lmFit(v, design)
  x = sapply(colnames(design),function(x)paste(x,'-(',paste(colnames(design)[colnames(design)!=x],collapse = "+"),')/4',sep = ''))
  con = makeContrasts(contrasts = x,levels = colnames(design))
  fit = contrasts.fit(fit,con)
  fit <- eBayes(fit, robust=TRUE)
  
  return(list(design = design, fit = fit))
}

createMarkers <- function( input, n_markers = 10){
  markers = NULL
  
  for(i in 1:ncol(input$design)){
    x = input$fit$coef[p.adjust(input$fit$p.v[,i],'fdr')<0.05,i]
    markers[[colnames(input$design)[i]]] =  names(head(sort(-x[x>0]),n_markers))
  }
  unlist(lapply(markers,length))
  
  full_genes = NULL
  for(i in 1:ncol(input$design)){
    x = input$fit$coef[p.adjust(input$fit$p.v[,i],'fdr')<0.05,i]
   full_genes[[colnames(input$design)[i]]] =  names(sort(-x[x>0]))
  }
  unlist(lapply(full_genes,length))
  
  return( list(markers = markers, full = full_genes) )
}


#createMarkers 
darmanis_fit <- compareReferenceCells(bulk_expression = geneExpr_our, reference_expression = Darmanis, reference_meta = DarmanisCells)

darmanis_markers <- createMarkers(darmanis_fit, n_markers = 100)
#darmanis <- darmanis_markers$markers
#save(darmanis, file = here::here("data/markers/Darmanis_single_cell.RData"))
```

Created set of Darmanis marker genes (100 per cell with highest fold change >0) and total genes.


## Deconvolution 

### dtangle with Darmanis reference 
```{r deconv.darmanis, echo=FALSE,warning=FALSE,messsage=FALSE}

# make normalising gene counts optional
# alter the number of markers
#darmanis_dtangle <- function(normalise = TRUE, n_markers)

summariseSingleCell <- function(expression_matrix, cell_meta){
  sapply(unique(DarmanisCells),function(x)rowMeans(Darmanis[,names(which(DarmanisCells==x))]))
}

#Darmanis

d_mean <- split(darmanis_metadata, darmanis_metadata$cellType) %>%
  map("cellID") %>%
  map_dfc( ~{
    rowMeans(cpm(Darmanis[,.x]))
  }) %>% as.data.frame()
row.names(d_mean) <- row.names(Darmanis)

# # old code
# y <- DGEList(counts= Darmanis)#[int,])
# y <- calcNormFactors(y,method = 'TMM', Acutoff =  quantile(log2(Darmanis[,1]/sum(Darmanis[,1])),0.75))
# DarmanisCPM <- cpm(y, log=FALSE)
# DarmanisMean = sapply(unique(DarmanisCells),function(x)rowMeans(DarmanisCPM[,names(which(DarmanisCells==x))]))
# 
# 

#dtangle <- 
sce <- d_mean #DarmanisMean #d_mean 
ge <- geneExpr_our
commongenes <- intersect (rownames(ge), rownames(sce))
ge <- ge[pmatch(commongenes, rownames(ge)), ]
sce <- sce[pmatch(commongenes, rownames(sce)), ]
y <- cbind(sce, ge)
y <- DGEList(counts= y)
y <- calcNormFactors(y,method = 'TMM', Acutoff =  quantile(log2(Darmanis[,1]/sum(Darmanis[,1])),0.75))
y <- cpm(y, log=FALSE)
y <- normalizeQuantiles(y)
pure_samples = as.list(1:5)
names(pure_samples) = colnames(y)[1:5]
markers = lapply(darmanis_markers$markers,intersect,rownames(y))
dtDarmanis <- dtangle(log2(t(y)), pure_samples=pure_samples, markers = markers[names(pure_samples)])$estimates


#write.table(dtDarmanis, file = paste0(plots_folder, "dtangleDarmanis_als.txt"), sep = "\t", quote = F)
```

### Darmanis - dtangle
The expression data comes from our dataset. 

```{r darmanis_dt, echo=FALSE,warning=FALSE,messsage=FALSE, fig.width=12, fig.height=20, res=300}

dtDarmanis_m <- melt(dtDarmanis)
colnames(dtDarmanis_m) <- c("sample", "cell", "deconv")
#head(dtDarmanis_m)
dtDarmanis_m$sample <- as.factor(dtDarmanis_m$sample)
metadata_dtangle_darmanis <- merge(dtDarmanis_m, metadata, by.x = "sample", by.y = "sample") # Both tables are in the same order 


darmanis_dtangle_res <- metadata_dtangle_darmanis


```

### Darmanis - MuSiC

```{r ,echo=FALSE,warning=FALSE,messsage=FALSE, fig.width=16, fig.height=20, res=300}
darmanis_music_res_file <- here::here("data/deconvolution/NYGC_Darmanis_MuSiC_deconv.RData")

if(rerun){
  # T is bulk; C is reference
  bulk <- geneExpr_our
  reference <- Darmanis[, darmanis_metadata$cellID]
  reference_meta <- as.data.frame(darmanis_metadata)
  sample_column = grep("[S-s]ample|[S-s]ubject",colnames(reference_meta))
  colnames(reference_meta)[sample_column] = "SubjectName"
  rownames(reference_meta) = reference_meta$cellID
  
  C.eset <- Biobase::ExpressionSet(assayData = as.matrix(reference),phenoData = Biobase::AnnotatedDataFrame(reference_meta))
  
  T.eset <- Biobase::ExpressionSet(assayData = as.matrix(bulk))
  
  # devtools::install_github('xuranw/MuSiC')
  library(MuSiC)
  require(xbioc)
  darmanis_music_res_df = t(music_prop(bulk.eset = T.eset, sc.eset = C.eset, clusters = 'cellType',
                                       markers = NULL, normalize = FALSE, samples = 'SubjectName', 
                                       verbose = TRUE)$Est.prop.weighted)
  
  darmanis_music_res <- t(darmanis_music_res_df) %>%
    as.data.frame() %>%
    rownames_to_column(var = "sample") %>%
    tidyr::gather(key = "cell", value = "deconv", -sample) %>%
    left_join(metadata, by = "sample")
  
  save(darmanis_music_res, file = darmanis_music_res_file)
  
}else{
  load(darmanis_music_res_file)
}


```

### Darmanis - DSA
The expression data comes from our dataset. 
DSA prefers small numbers of markers so take top 10

```{r , echo=FALSE,warning=FALSE,messsage=FALSE,fig.width=20, fig.height=8, res=300}
# have to download from web
# code below assumes you'll download to current working directory
# change this to something different
#install.packages("BiocInstaller", repos="http://bioconductor.org/packages/3.7/bioc/")
# XML - install legacy binary
 #install.packages("https://cran.r-project.org/bin/macosx/el-capitan/contrib/3.6/XML_3.99-0.3.tgz" )
#install.packages("csSAM")
# download this: http://web.cbio.uct.ac.za/~renaud/CRAN/src/contrib/CellMix_1.6.2.tar.gz 
#system("R CMD INSTALL CellMix_1.6.2.tar.gz")
#require(CellMix)

#BiocManager::install(c("GSEABase", "annotate", "genefilter", "preprocessCore", "graph") )
#marker_distrib is output of marker_strategies(topTable_RESULTS, marker_strategy, C)
# 
# darmanis_markers10 <- createMarkers(darmanis_fit, n_markers = 10)
# 
# #DarmanisMarkers
# bulk <- geneExpr_our
# ML <- CellMix::MarkerList(darmanis_markers10$markers)
# #ML@.Data <- tapply(as.character(md$gene),as.character(md$CT),list)
# darmanis_dsa_res_full <- CellMix::ged(as.matrix(bulk), ML, method = "DSA", log = FALSE)@fit@H
# darmanis_dsa_res_full <- apply(darmanis_dsa_res_full,2,function(x) ifelse(x < 0, 0, x)) #explicit non-negativity constraint
# darmanis_dsa_res_full <- apply(darmanis_dsa_res_full,2,function(x) x/sum(x)) #explicit STO constraint
# 
# darmanis_dsa_res <- t(darmanis_dsa_res_full) %>%
#   as.data.frame() %>%
#   rownames_to_column(var = "sample") %>%
#   tidyr::gather(key = "cell", value = "deconv", -sample) %>%
#   left_join(metadata, by = "sample")
```

## Mathys et al Human Brain 

## dtangle

Use markers created by Rahat from the Mathys cell types

Also need the d_mean object, the mean expression in each cell type, to add.

```{r}
load("data/deconvolution/Mathys_markers.RData")
load("data/deconvolution/Mathys_d_mean.RData")
sce <- d_mean #DarmanisMean #d_mean 
ge <- geneExpr_our
commongenes <- intersect (rownames(ge), rownames(sce))
ge <- ge[pmatch(commongenes, rownames(ge)), ]
sce <- sce[pmatch(commongenes, rownames(sce)), ]
y <- cbind(sce, ge)
y <- DGEList(counts= y)
y <- calcNormFactors(y,method = 'TMM', Acutoff =  quantile(log2(Darmanis[,1]/sum(Darmanis[,1])),0.75))
y <- cpm(y, log=FALSE)
y <- normalizeQuantiles(y)
pure_samples = as.list(1:8)
names(pure_samples) = colnames(y)[1:8]
markers = lapply(Mathys_markers$markers,intersect,rownames(y))
dtMathys <- dtangle(log2(t(y)), pure_samples=pure_samples, markers = markers[names(pure_samples)])$estimates

dtMathys_m <- melt(dtMathys)
colnames(dtMathys_m) <- c("sample", "cell", "deconv")
#head(dtMathys_m)
dtMathys_m$sample <- as.factor(dtMathys_m$sample)
metadata_dtangle_Mathys <- merge(dtMathys_m, metadata, by.x = "sample", by.y = "sample") # Both tables are in the same order 

mathys_dtangle_res <- metadata_dtangle_Mathys



```




### MuSiC
 
### Create markers
 
```{r}
# library(readr)
# library(here)
# library(dplyr)
# 
# load(here::here("data/feb_2020/gene_counts_no_tags.RData") )
# 
# metadata <- read_tsv( here::here("data/differential_expression/ALS/ALS_all_differential_expression_support.tsv") )
# 
# # filter out non-CNS samples
# tissues <- c("Lumbar_Spinal_Cord", "Cervical_Spinal_Cord", "Thoracic_Spinal_Cord")
# 
# # remove low RIN samples - for good
# metadata <- filter(metadata, tissue %in% tissues)
# 
# #Gene name for the voom table 
# gencode_30 <- read_tsv(here::here("data/misc/gencode.v30.tx2gene.tsv.gz")) %>%
#   janitor::clean_names() %>%
#   dplyr::select(ensembl = geneid, symbol = genename) %>%
#   dplyr::distinct() %>%
#   dplyr::mutate(ensembl =  str_split_fixed(ensembl, "\\.", 2)[,1] ) %>%
#   dplyr::distinct() 
# 
# genes_loc <- as.data.frame(genes_counts[, metadata$sample])
# 
# # remove lowly expressed genes
# keep.exp <- rowSums(cpm(genes_loc) > 1) >= ceiling(0.5*ncol(genes_loc))
# genes_loc <- genes_loc[keep.exp,]
# 
# geneExpr_our = as.data.frame(genes_loc)
# geneExpr_our$ensembl = row.names(geneExpr_our)
# geneExpr_our <- merge(geneExpr_our, gencode_30, by = "ensembl")
# #Remove duplicates 
# geneExpr_our <- geneExpr_our[! duplicated(geneExpr_our$symbol),]
# dim(geneExpr_our) # 19730   276
# rownames(geneExpr_our) = geneExpr_our$symbol
# geneExpr_our$ensembl = NULL
# geneExpr_our$symbol = NULL
# 
# 
# 
# # differential expression - compare each cell type to the mean
# compareReferenceCells <- function(bulk_expression, reference_expression, reference_meta){
#   # only use reference genes present in bulk
#   int <- intersect(rownames(bulk_expression),rownames(reference_expression))
#   y <- DGEList(counts= reference_expression[int,])
#   y <- calcNormFactors(y,method = 'TMM', Acutoff =  quantile(log2(reference_expression[,1]/sum(reference_expression[,1])),0.75))
#   
#   reference_expression <- cpm(y, log=FALSE)
#   reference_expressionMean = sapply(unique(reference_meta),function(x)rowMeans(reference_expression[,names(which(reference_meta==x))]))
#   int = intersect(rownames(bulk_expression),rownames(reference_expressionMean))
#   design = model.matrix(~.-1,data.frame(cell= reference_meta[colnames(reference_expression)]))
#   colnames(design) = gsub('cell','',colnames(design))
#   v <- voom(reference_expression[int,rownames(design)], design, plot=FALSE)
#   fit <- lmFit(v, design)
#   x = sapply(colnames(design),function(x)paste(x,'-(',paste(colnames(design)[colnames(design)!=x],collapse = "+"),')/4',sep = ''))
#   con = makeContrasts(contrasts = x,levels = colnames(design))
#   fit = contrasts.fit(fit,con)
#   fit <- eBayes(fit, robust=TRUE)
#   
#   return(list(design = design, fit = fit))
# }
# 
# createMarkers <- function( input, n_markers = 10){
#   markers = NULL
#   
#   for(i in 1:ncol(input$design)){
#     x = input$fit$coef[p.adjust(input$fit$p.v[,i],'fdr')<0.05,i]
#     markers[[colnames(input$design)[i]]] =  names(head(sort(-x[x>0]),n_markers))
#   }
#   unlist(lapply(markers,length))
#   
#   full_genes = NULL
#   for(i in 1:ncol(input$design)){
#     x = input$fit$coef[p.adjust(input$fit$p.v[,i],'fdr')<0.05,i]
#    full_genes[[colnames(input$design)[i]]] =  names(sort(-x[x>0]))
#   }
#   unlist(lapply(full_genes,length))
#   
#   return( list(markers = markers, full = full_genes) )
# }
# 
# 
# ## readMM requires the Matrix package
# mathys_counts <- Matrix::readMM(here::here("data/Mathys_Brain_scRNA/filtered_count_matrix.mtx.gz") )
# rownames(mathys_counts) <- readLines(here::here("data/Mathys_Brain_scRNA/filtered_gene_row_names.txt") )
mathys_meta <- read_tsv(here::here("data/Mathys_Brain_scRNA/filtered_column_metadata.txt") ) %>%
   janitor::clean_names()
#colnames(mathys_counts) <- mathys_meta$tag
# 
 # projid is subject I think
# #length(table(mathys_meta$projid))
# table(mathys_meta$broad_cell_type)
# 
mathys_meta <- dplyr::select(mathys_meta, cellID = tag, cellType = broad_cell_type, sampleID = projid)
# 
# # load in ROSMAP metadata
# # keep only controls
rosmap_meta <- read_csv(here::here("data/Mathys_Brain_scRNA/ROSMAP_Clinical_2019-05_v3.csv")) %>%
   filter(projid %in% mathys_meta$sampleID, cogdx == 1)
# 
# #keep only control individuals
# mathys_controls_metadata <- filter(mathys_meta, sampleID %in% rosmap_meta$projid )
# 
# mathys_control_cells <- mathys_controls_metadata$cellType
# names(mathys_control_cells) <- mathys_controls_metadata$cellID
# # 23,513 cells left
# mathys_controls_counts <- mathys_counts[, colnames(mathys_counts) %in% mathys_controls_metadata$sampleID] 
# 
# write(mathys_controls_metadata, mathys_controls_counts, file = "/data/Mathys_Brain_scRNA/Mathys_controls.RData")
# 
# #createMarkers 
# mathys_fit <- compareReferenceCells(bulk_expression = geneExpr_our, reference_expression = mathys_controls_counts, reference_meta = mathys_control_cells)
# 
# mathys_markers <- createMarkers(darmanis_fit, n_markers = 100)
# 
# write(mathys_fit, mathys_markers, file = "/data/Mathys_Brain_scRNA/Mathys_markers.RData")

```



```{r, echo = FALSE, warn = FALSE, message = FALSE}
## do deconvolution using Mathys single cell as reference

# T is bulk; C is reference

## test first on just temporal cortex
#tcx_only <- metadata_filt[metadata_filt$tissue == "Temporal_Cortex",]
#tcx_only_counts <- geneExpr_our[, tcx_only$sample,]

mathys_music_file <- here::here("data/deconvolution/NYGC_Mathys_deconv.RData")

if( rerun  ){
  ##install.packages(Matrix)
  ## readMM requires the Matrix package
  mathys_counts <- Matrix::readMM(here::here("data/Mathys_Brain_scRNA/filtered_count_matrix.mtx.gz") )
  rownames(mathys_counts) <- readLines(here::here("data/Mathys_Brain_scRNA/filtered_gene_row_names.txt") )
  
  mathys_meta <- read_tsv(here::here("data/Mathys_Brain_scRNA/filtered_column_metadata.txt") ) %>%
    janitor::clean_names() %>%
    dplyr::select(cellID = tag, cellType = broad_cell_type, sampleID = projid)
  
  colnames(mathys_counts) <- mathys_meta$cellID
  
  
  rosmap_meta <- read_csv(here::here("data/Mathys_Brain_scRNA/ROSMAP_Clinical_2019-05_v3.csv")) %>%
    filter(projid %in% mathys_meta$sampleID, cogdx == 1)
  
  #keep only control individuals
  mathys_controls_metadata <- filter(mathys_meta, sampleID %in% rosmap_meta$projid )

  mathys_control_cells <- mathys_controls_metadata$cellType
  names(mathys_control_cells) <- mathys_controls_metadata$cellID
  # 23,513 cells left

  mathys_controls_counts <- mathys_counts[, colnames(mathys_counts) %in% mathys_controls_metadata$cellID]

  # # for Jake - random sample 10% of counts
  # mathys_subset <- mathys_counts[, sample(1:ncol(mathys_counts), size = 100 ) ]
  # mathys_meta_subset <- dplyr::filter(mathys_meta, tag %in% colnames(mathys_subset))
  # 
  # mathys_meta_subset <- dplyr::select(mathys_meta_subset, cell_id = tag, everything() )
  # 
  # projid is subject I think
  #length(table(mathys_meta$projid))
  #table(mathys_meta$broad_cell_type)
  
  

  # add together Ex and In to have only one neuronal class:
  # mathys_meta_one_neuron <- mathys_meta %>%
  #   mutate(cellType = case_when( cellType %in% c("In","Ex") ~ "Neuron", TRUE ~ cellType ) )
  # 
  # table(mathys_meta_one_neuron$cellType)
  # 
  #23,513 cells
  bulk <- geneExpr_our
  #bulk <- tcx_only_counts
  reference <- mathys_controls_counts
  reference_meta <- as.data.frame( mathys_controls_metadata, stringsAsFactors = FALSE)
  
  sample_column = grep("[S-s]ample|[S-s]ubject",colnames(reference_meta))
  colnames(reference_meta)[sample_column] = "SubjectName"
  rownames(reference_meta) = reference_meta$cellID
  
  C.eset <- Biobase::ExpressionSet(assayData = as.matrix(reference),phenoData = Biobase::AnnotatedDataFrame(reference_meta))
  
  T.eset <- Biobase::ExpressionSet(assayData = as.matrix(bulk))
  
  #library(MuSiC)
  require(xbioc)
  library(MuSiC)
  mathys_music_results <- music_prop(bulk.eset = T.eset, sc.eset = C.eset, clusters = 'cellType',
                                     markers = NULL, normalize = FALSE, samples = 'SubjectName', 
                                     verbose = TRUE)$Est.prop.weighted
  
  mathys_music_res <- mathys_music_results %>%
    as.data.frame() %>%
    rownames_to_column(var = "sample") %>%
    tidyr::gather(key = "cell", value = "deconv", -sample) %>%
    left_join(metadata, by = "sample")
  
  save(mathys_music_res, file = mathys_music_file )
}else{
  load(mathys_music_file)
}

dim(mathys_music_res)


```


# Comparison between methods and reference sets



## Regress out technical and other factors and compare deconvolution estimates between ALS and Controls

```{r fig.width = 12, fig.height = 8}


rna_tech_df <- read_tsv(here::here("data/differential_expression/ALS/ALS_all_differential_expression_tech_metrics.tsv"))

residualiser <- function(data){
  data <- left_join(data, rna_tech_df, by = "sample") 
  if("onset" %in% names(data)){
    data$onset <- NULL
  }
  data <- data[complete.cases(data),]
  #print(table(data$prep))
  mod <- lm( deconv ~ factor(prep) + age + factor(sex) + rin + pct_mrna_bases + pct_ribosomal_bases + pct_chimeras + pct_intergenic_bases + median_3prime_bias + median_5prime_bias, data)
  data$resid <-  residuals(mod) 
  
  data$p_nom <- pairwise.wilcox.test(x = data$deconv, g = data$disease)$p.value[1]
  data$p_resid <- pairwise.wilcox.test(x = data$resid, g = data$disease)$p.value[1]
  
  return(data)
}

```


```{r fig.width = 8, fig.height = 5}
#darmanis_dsa_resid <- darmanis_dsa_res %>% filter(sample != "CGND-HRA-00431.1" ) %>% split(paste(.$tissue, .$cell) ) %>% map_df( residualiser)
if(rerun){
darmanis_dtangle_resid <- darmanis_dtangle_res %>% filter(sample != "CGND-HRA-00431.1" ) %>% split(paste(.$tissue, .$cell) ) %>% map_df( residualiser)

darmanis_music_resid <- darmanis_music_res %>% filter(sample != "CGND-HRA-00431.1" ) %>% split(paste(.$tissue, .$cell) ) %>% map_df( residualiser)

mathys_music_resid <- mathys_music_res %>% filter(sample != "CGND-HRA-00431.1" ) %>% split(paste(.$tissue, .$cell) ) %>% map_df( residualiser)

mathys_dtangle_resid <- mathys_dtangle_res %>% filter(sample != "CGND-HRA-00431.1" ) %>% split(paste(.$tissue, .$cell) ) %>% map_df( residualiser)


#write_tsv(darmanis_dsa_resid, file = here::here("data/deconvolution/Darmanis_DSA_residual_results.tsv") )
write_tsv(darmanis_music_resid, file = here::here("data/deconvolution/Darmanis_MuSiC_residual_results.tsv") )
write_tsv(darmanis_dtangle_resid, file = here::here("data/deconvolution/Darmanis_dtangle_residual_results.tsv") )
write_tsv(mathys_music_resid, file = here::here("data/deconvolution/Mathys_MuSiC_residual_results.tsv") )
write_tsv(mathys_dtangle_resid, file = here::here("data/deconvolution/Mathys_dtangle_residual_results.tsv") )
}else{

darmanis_music_resid <- read_tsv(here::here("data/deconvolution/Darmanis_MuSiC_residual_results.tsv") )
darmanis_dtangle_resid <- read_tsv( here::here("data/deconvolution/Darmanis_dtangle_residual_results.tsv") )
mathys_music_resid <- read_tsv(here::here("data/deconvolution/Mathys_MuSiC_residual_results.tsv"))
mathys_dtangle_resid  <- read_tsv(here::here("data/deconvolution/Mathys_dtangle_residual_results.tsv"))

}
## write out MuSiC estimates with Mathys cells for supplementary table
mathys_music_df <-
  mathys_music_resid %>%
  dplyr::select(sample, cell, deconv) %>%
  pivot_wider(names_from = cell, values_from = deconv)

write_csv(mathys_music_df, file = here::here("ALS_SC/tables/mathys_music_deconvolution_table.csv"))



```


## Plotting

```{r fig.width = 7, fig.height = 3}

estimate_plot <- function(
  results, p_df, method, reference = "Darmanis", ycol = "deconv", p = "p_nom", 
  tissues = c("Cervical_Spinal_Cord", "Lumbar_Spinal_Cord", "Thoracic_Spinal_Cord")  
  ){
  
  to_plot <- results %>%
    mutate(tissue = factor(tissue, levels = tissues)) %>%
    filter(tissue %in% tissues) %>%
    filter(disease %in% c("ALS","Control")) %>%
    mutate( disease = forcats::fct_relevel(disease,"Control")) %>%
    mutate(tissue = factor( gsub("_","\n", tissue), levels = gsub("_","\n", tissues) )  ) %>%
    filter( !is.na(cell))
  
  plot <- 
  to_plot %>%
  ggplot(aes(x = disease, deconv)) +
    rasterise(
      geom_point(aes(colour = disease),size = 0.5, position = position_jitter(width = 0.33, height = 0) ),
      dpi = 300
      )+
    geom_boxplot(fill = NA,notch = FALSE, na.rm = TRUE, outlier.color = NA) +
    facet_grid(tissue ~ cell,switch = "y" ) +
    scale_colour_manual(values = c("#4F8FC4", "#B61927")) +
    labs(subtitle = paste0( reference, " - ", method) , x = "", y = "" ) +
    guides(fill = FALSE, colour = FALSE) +
    theme_jh() +
    theme(
      plot.title = element_blank(),
      strip.background = element_blank(),
      strip.switch.pad.grid = unit(0,units ="pt"),
      panel.spacing.x = unit(0,units = "pt"), 
      strip.placement = "outside",
      panel.border = element_blank(),
      strip.text.x = element_text(face = "bold", hjust = 0.5, vjust = 1, margin = margin(t =5, b = 5, unit = "pt")),
      strip.text.y.left = element_text(angle = 0),
      plot.margin = margin()
    ) +
    labs(y = "") +
    scale_y_continuous(labels = scales::percent_format(accuracy = 1), limits = c(0,1)) + 
    geom_text(nudge_x = 0.5, data = p_df, aes(x = 1, y = 0.9, label = padj_label)   )
  
  return(plot)
}

conversion_table <- data.frame(short = c("Ast","End", "Mic", "Ex", "Oli", "Per" ),
                               long = c("astrocytes", "endothelial", "microglia", "neurons", "oligodendrocytes", "pericytes"),
                               mono = c("A","E","M","N","O","P"),
                               title = c("Astrocytes", "Endothelial", "Microglia", "Neurons", "Oligos", "Pericytes")
  )

mathys_music_df <- mathys_music_resid %>%
  left_join(conversion_table, by = c("cell" = "short") ) %>%
  select(-cell) %>%
  rename(cell = long) %>%
  select(sample, disease, duration, cell, tissue, deconv, resid, p_nom, p_resid)

mathys_dtangle_df <- mathys_dtangle_resid %>%
  left_join(conversion_table, by = c("cell" = "short") ) %>%
  select(-cell) %>%
  rename(cell = long) %>%
  select(sample, disease, duration, cell, tissue, deconv, resid, p_nom, p_resid)

# each tool / reference comparison needs Bonferroni adjusting of the P-values
get_bonf_res <- function(data){
  data %>% select(cell, tissue, p_nom, p_resid) %>%
    distinct() %>%
    drop_na() %>%
    mutate(tissue = gsub("_","\n", tissue) )  %>% 
    mutate(q_nom = p.adjust(p_nom, method = "bonferroni"),
           q_resid = p.adjust(p_resid, method = "bonferroni")
           ) %>%
    mutate(padj_label = case_when(
    q_resid < 0.0001 ~ "***",
    q_resid < 0.001 ~ "**",
    q_resid < 0.05 ~ "*",
    TRUE ~ "."
    ))
}

mathys_music_q <- get_bonf_res(mathys_music_df)
mathys_dtangle_q <- get_bonf_res(mathys_dtangle_df)

darmanis_music_q <- get_bonf_res(darmanis_music_resid)
darmanis_dtangle_q <- get_bonf_res(darmanis_dtangle_resid)

# plot estimates

mathys_music_csc <- estimate_plot(mathys_music_df, mathys_music_q, method = "MuSiC", reference = "Mathys single nucleus", tissues = c("Cervical_Spinal_Cord"))

mathys_music_csc



```

Main figure - show Cervical Spinal Cord with MuSiC estimates

```{r}

main_plot_q <- mathys_music_q %>% left_join(conversion_table, by = c("cell" = "long")) %>%
  filter(tissue == "Cervical\nSpinal\nCord")

main_csc_plot <- 
  mathys_music_df %>%
  left_join(conversion_table, by = c("cell" = "long") ) %>%
  drop_na() %>%
  filter(tissue == "Cervical_Spinal_Cord") %>%
  mutate(disease = factor(disease, levels = c("Control", "ALS"))) %>%
  mutate(disease_group = ifelse(disease == "Control", "A", "B")) %>%
  mutate(box_group = paste(disease_group,title ) ) %>%
  ggplot(aes(x = title, deconv)) +
    rasterise(
      geom_point(aes(colour = disease),
                 size = 0.5, 
                 position = position_jitterdodge(dodge.width = 0.8, 
                                                 jitter.width = 0.5, jitter.height = 0) ),
      dpi = 300)+
    geom_boxplot(aes(group = box_group), 
                 colour = "black", fill = NA,notch = FALSE, na.rm = TRUE, outlier.color = NA, 
                 position = position_dodge(width = 0.8)
                 #, 
                 #width = 0.5
                 ) +
    scale_colour_manual(name = "", values = c("#4F8FC4", "#B61927")) +
    labs(x = "", y = "" ) +
    guides(fill = FALSE) +
    theme_jh() +
    theme(
      axis.ticks = element_line(colour = "black"),
      legend.position = "top",
      axis.text.x = element_text(angle = 45, hjust = 1, size = 7),
      plot.title = element_blank(),
      #strip.background = element_blank(),
      #strip.switch.pad.grid = unit(0,units ="pt"),
      #panel.spacing.x = unit(0,units = "pt"), 
      #strip.placement = "outside",
      panel.border = element_blank(),
      #strip.text.x = element_text(hjust = 0.5, vjust = 1, margin = margin(t =5, b = 5, unit = "pt")),
      #strip.text.y.left = element_text(angle = 0),
      #plot.margin = margin()
    ) +
    labs(y = "") +
    geom_text( 
              data = main_plot_q, 
              aes(x = title , y = 0.85, label = padj_label)   ) +
  scale_y_continuous(name = "Estimated proportion", labels = scales::percent_format(accuracy = 1), limits = c(0,0.87)) 

main_csc_plot

ggsave(plot = main_csc_plot, filename = here::here("ALS_SC/plots/main_figure_mathys_music_csc_deconvolution_plot.pdf"), width = 3.5, height = 2.75)



```



## Plot all MuSiC estimates - for supplement

```{r fig.height = 12, fig.width = 8}
music_multiplot <- 
  estimate_plot(mathys_music_df, method = "MuSiC", reference = "Mathys single nucleus") +
  estimate_plot(darmanis_music_resid, method = "MuSiC", reference = "Darmanis single cell") +
    plot_layout(ncol = 1) +
  plot_annotation(tag_levels = "A") &
  theme(plot.tag = element_text(face = "bold"))

ggsave(plot = music_multiplot, filename =  here::here("ALS_SC/plots/deconvolution_music_plots.pdf"), width = 8, height = 9)
ggsave(plot = music_multiplot, filename =  here::here("ALS_SC/plots/deconvolution_music_plots.png"), width = 8, height = 9, dpi = 600)

music_multiplot
```


## All dtangle estimates 

```{r fig.height = 4.5, fig.width = 8}
dtangle_multiplot <- 
  estimate_plot(mathys_dtangle_df, method = "dtangle", reference = "Mathys single nucleus") +
    estimate_plot(darmanis_dtangle_resid, method = "dtangle", reference = "Darmanis single cell") +
  theme(plot.margin = margin(t = 5,b=5,l=5,r =5, unit = "pt")) +
  plot_layout(ncol = 1) +
  plot_annotation(tag_levels = "A") &
  theme(plot.tag = element_text(face = "bold"))

dtangle_multiplot
ggsave(plot = dtangle_multiplot, filename =  here::here("ALS_SC/plots/deconvolution_dtangle_plots.pdf"), width = 8, height = 9)
ggsave(plot = dtangle_multiplot, filename =  here::here("ALS_SC/plots/deconvolution_dtangle_plots.png"), width = 8, height = 9, dpi = 600)

```


## NEW for Revisions - compare marker gene lists between Darmanis and Mathys

Finish!

```{r}
darmanis_mathys_conversion <- data.frame(
  mathys = c("Ast", "End", "Ex", "In", "Mic", "Oli", "Opc", "Per"),
  darmanis = c("astrocytes", "endothelial", "neurons", NA, "microglia", "oligodendrocytes", NA, NA)
)



d_mark <- darmanis_markers$markers
m_mark <- Mathys_markers$markers
names(m_mark) <- darmanis_mathys_conversion$darmanis[ match(names(m_mark), darmanis_mathys_conversion$mathys)]

m_mark <- m_mark[ !is.na(names(m_mark))]

#names(d_mark) <- paste0("d_", names(d_mark) )
#names(m_mark) <- paste0("m_", names(m_mark))
#all_mark <- c(d_mark, m_mark) 

#upset(fromList( all_mark  ), nsets = length(all_mark))

upset(fromList( d_mark  ), nsets = length(d_mark))
upset(fromList( m_mark  ), nsets = length(m_mark))

#jaccard_intersect <- function(x,y){ length(intersect(x,y) ) / length(unique(c(x,y))) }

#jaccard_intersect(d_mark$d_astrocytes, m_mark$m_astrocytes)

#mark_expand <- crossing(Var1 = names(d_mark), Var2 = names(m_mark) )

mark_jac_res <- map2_df(names(d_mark),names(d_mark), ~{
    int <- length(intersect(d_mark[[.x]], m_mark[[.y]]) ) 
    uni <- length(unique(c(d_mark[[.x]], m_mark[[.y]]) ) ) 

    jac <- int / uni
    
    tibble(x = .x, y = .y, int = int, uni = uni, jac = jac)
  })

mark_jac_res %>%
  mutate(jac = ifelse(jac == 1, NA, jac)) %>%
  ggplot(aes(x, y, fill = jac)) + geom_tile() + geom_text(aes(label = int) )

mark_jac_res
  

```



For pairs of results, 
calculate correlation between the estimate of neuronal, astrocyte and microglial proportion in each sample

```{r fig.width = 8, fig.height = 5}
compare_2_results <- function(res1, res2, res1.name, res2.name){
  # conversion_table <- data.frame(short = c("astro","endo", "microglia", "neuro", "oligo" ),
  #                              long = c("astrocytes", "endothelial", "microglia", "neurons", "oligodendrocytes")
  # )
  # # match ids
  # if( any(res1$cell %in% conversion_table$short) ){
  #   res1$cell <- conversion_table$long[ match(res1$cell, conversion_table$short)]
  # }
  # if( all(res2$cell %in% conversion_table$short) ){
  #   res2$cell <- conversion_table$long[ match(res2$cell, conversion_table$short)]
  # }
  
  left_join(
  x = dplyr::select(res1, sample, cell, deconv),
  y =dplyr::select(res2, sample, cell, deconv), 
  suffix = c(".res1", ".res2"),
  by = c("sample", "cell")
) %>%
  left_join(metadata, by = "sample") %>%
    mutate(tissue = gsub("_", "\n", tissue)) %>%
    filter(!is.na(tissue)) %>%
    filter(!is.na(cell)) %>%
    #filter(tissue %in% c("Lumbar_Spinal_Cord", "Cervical_Spinal_Cord")) %>%
  ggplot(aes(x = deconv.res1, y = deconv.res2 ) ) + 
    geom_point( size = 0.5, aes(colour = disease), alpha = 0.5) +
    geom_abline(slope =1 , intercept =0, linetype = 3) +
    labs(x = res1.name, y = res2.name)+
    theme_jh() +
    facet_grid(tissue ~ cell,switch = "y") + #, scales = "free" ) +
    ggpubr::stat_cor(method = "spearman",aes(label = ..r.label..), size = 3  ) +
    scale_colour_manual(values = c("red", "darkblue")) +
    theme(
      #plot.title = element_blank(),
      strip.background = element_blank(),
      strip.switch.pad.grid = unit(0,units ="pt"),
      panel.spacing.x = unit(10,units = "pt"), 
      panel.spacing.y = unit(10,units = "pt"), 
      strip.placement = "outside",
      panel.border = element_blank(),
      strip.text.x = element_text(face = "bold", hjust = 0.5, vjust = 1, margin = margin(t =5, b = 5, unit = "pt")),
      strip.text.y.left = element_text(angle = 0),
    ) +
    scale_y_continuous(labels = scales::percent_format(accuracy = 1),n.breaks = 3, breaks = waiver() ) +
    scale_x_continuous(labels = scales::percent_format(accuracy = 1),n.breaks = 3, breaks = waiver() ) 
}


darmanis_compare_plot <- 
  compare_2_results(res1 = darmanis_music_res, res2 = darmanis_dtangle_res, "MuSiC", "dtangle") + labs(title = "Darmanis - MuSiC vs dtangle", x = "MuSiC estimate", y = "dtangle estimate") #+

mathys_compare_plot <- 
  compare_2_results(mathys_music_df, mathys_dtangle_df, "MuSiC", "dtangle") + labs(title = "Mathys - MuSiC vs dtangle", x = "MuSiC estimate", y = "dtangle estimate") #+


reference_compare_plot <- mathys_compare_plot + darmanis_compare_plot + plot_layout(ncol = 1)

ggsave(plot = reference_compare_plot, filename =  here::here("ALS_SC/plots/deconvolution_reference_compare_plot.pdf"), width = 10, height = 8)



```

Endothelial cells have worst agreement between MuSiC and the two marker based methods dtangle and DSA. 
Should I remove endothelial cells as a category and rerun?


Compare MuSiC results from Darmanis and Mathys single cell data
```{r fig.width = 9, fig.height = 6}
#table(msc_music_res$cell)
darmanis_mathys_conversion <- data.frame(
  mathys = c("Ast", "End", "Ex", "In", "Mic", "Oli", "Opc", "Per"),
  darmanis = c("astrocytes", "endothelial", "neurons", "neurons", "microglia", "oligodendrocytes", NA, NA)
)

mathys_compare <- mathys_music_res %>% dplyr::select(sample, cell, deconv)
mathys_compare$cell_mathys <- mathys_compare$cell

mathys_compare$cell <- darmanis_mathys_conversion$darmanis[match(mathys_compare$cell, darmanis_mathys_conversion$mathys)]


music_reference_compare_plot <-
  left_join(mathys_compare, 
            dplyr::select(darmanis_music_res, sample, cell, deconv), by = c("sample", "cell"), 
            suffix = c(".mathys", ".darmanis")) %>%
  left_join(metadata, by = "sample") %>%
  mutate(tissue = gsub("_", "\n", tissue )) %>%
  #filter(tissue %in% c("Cervical_Spinal_Cord", "Lumbar_Spinal_Cord")) %>%
  filter(cell_mathys %in% c("Ast", "End", "Ex", "Mic", "Oli")) %>%
  ggplot(aes(x = deconv.mathys, y = deconv.darmanis)) + 
    geom_point( size = 0.5, aes(colour = disease), alpha = 0.5) +
    geom_abline(slope =1 , intercept =0, linetype = 3) +
    #labs(x = res1.name, y = res2.name)+
    theme_jh() +
    facet_grid(tissue ~ cell,switch = "y") + #, scales = "free" ) +
    ggpubr::stat_cor(method = "spearman",aes(label = ..r.label..), size = 3 ) +
    scale_colour_manual(values = c("red", "darkblue")) +
    theme(
      strip.background = element_blank(),
      strip.switch.pad.grid = unit(0,units ="pt"),
      panel.spacing.x = unit(10,units = "pt"), 
      panel.spacing.y = unit(10,units = "pt"), 
      strip.placement = "outside",
      panel.border = element_blank(),
      strip.text.x = element_text(face = "bold", hjust = 0.5, vjust = 1, margin = margin(t =5, b = 5, unit = "pt")),
      strip.text.y.left = element_text(angle = 0),
      axis.text = element_text(size = 8)
    ) +
    scale_y_continuous(labels = scales::percent_format(accuracy = 1),n.breaks = 3, breaks = waiver() ) +
    scale_x_continuous(labels = scales::percent_format(accuracy = 1),n.breaks = 3, breaks = waiver() ) +
  labs(x = "MuSiC estimate with Mathys", y = "MuSiC estimate with Darmanis", title = "MuSiC - Mathys vs Darmanis")

#ggsave(plot = music_reference_compare_plot, filename =  here::here("ALS_SC/plots/deconvolution_music_reference_compare_plot.pdf"), width = 10, height = 6)


music_reference_compare_plot

## compare dtangle between references

mathys_compare <- mathys_dtangle_res %>% dplyr::select(sample, cell, deconv)
mathys_compare$cell_mathys <- mathys_compare$cell

mathys_compare$cell <- darmanis_mathys_conversion$darmanis[match(mathys_compare$cell, darmanis_mathys_conversion$mathys)]


dtangle_reference_compare_plot <-
  left_join(mathys_compare, 
            dplyr::select(darmanis_dtangle_res, sample, cell, deconv), by = c("sample", "cell"), 
            suffix = c(".mathys", ".darmanis")) %>%
  left_join(metadata, by = "sample") %>%
  mutate(tissue = gsub("_", "\n", tissue )) %>%
  #filter(tissue %in% c("Cervical_Spinal_Cord", "Lumbar_Spinal_Cord")) %>%
  filter(cell_mathys %in% c("Ast", "End", "Ex", "Mic", "Oli")) %>%
  ggplot(aes(x = deconv.mathys, y = deconv.darmanis)) + 
    geom_point( size = 0.5, aes(colour = disease), alpha = 0.5) +
    geom_abline(slope =1 , intercept =0, linetype = 3) +
    #labs(x = res1.name, y = res2.name)+
    theme_jh() +
    facet_grid(tissue ~ cell,switch = "y" ) +
    ggpubr::stat_cor(method = "spearman",aes(label = ..r.label..), size = 3 ) +
    scale_colour_manual(values = c("red", "darkblue")) +
    theme(
      strip.background = element_blank(),
      strip.switch.pad.grid = unit(0,units ="pt"),
      panel.spacing.x = unit(15,units = "pt"), 
      panel.spacing.y = unit(10,units = "pt"), 
      strip.placement = "outside",
      panel.border = element_blank(),
      strip.text.x = element_text(face = "bold", hjust = 0.5, vjust = 1, margin = margin(t =5, b = 5, unit = "pt")),
      strip.text.y.left = element_text(angle = 0),
      axis.text = element_text(size = 8)
    ) +
    scale_y_continuous(labels = scales::percent_format(accuracy = 1),n.breaks = 3, breaks = waiver() ) +
    scale_x_continuous(labels = scales::percent_format(accuracy = 1),n.breaks = 3, breaks = waiver() ) +
  labs(x = "dtangle estimate with Mathys", y = "dtangle estimate with Darmanis", title = "dtangle - Mathys vs Darmanis")


tool_compare_plot <- music_reference_compare_plot +
dtangle_reference_compare_plot + plot_layout(ncol = 1)


ggsave(plot = tool_compare_plot, filename =  here::here("ALS_SC/plots/deconvolution_tool_reference_compare_plot.pdf"), width = 10, height = 8)



```

Compare lumbar and cervical spinal cord estimates for shared donors

```{r}

all_res <- 
  bind_rows(
mathys_music_df %>% mutate(ref = "Mathys", tool = "MuSiC" ),
mathys_dtangle_df %>% mutate(ref = "Mathys", tool = "dtangle" ),
darmanis_music_res %>% mutate(ref = "Darmanis", tool = "MuSiC" ),
darmanis_dtangle_res %>% mutate(ref = "Darmanis", tool = "dtangle" )
) %>%
  mutate(donor = darmanis_music_res$donor[match(.$sample, darmanis_music_res$sample)]) %>%
  select(sample,donor, disease, cell, tissue, deconv, ref, tool) %>%
  drop_na() %>%
  mutate(cell = ifelse(cell == "oligos", "oligodendrocytes", cell))

all_res %>% pull(donor) %>% unique() %>% length()
# 203 unique donors with both CSP and LSP


## plot all four estimates as comparison
ref_comparison_boxplot <- 
  all_res %>%
  mutate(tissue = gsub("_", " ", tissue)) %>%
  ggplot(aes(y = deconv, x = cell)) +  
  geom_boxplot(aes(fill = paste(ref,tool) ), position = position_dodge(), outlier.size = 1 ) + 
  facet_wrap(~ tissue, ncol = 1) +
  theme_classic() +
  labs(fill = "", y = "Estimated proportion", x = "") +
      theme(
      strip.background = element_blank(),
      strip.switch.pad.grid = unit(0,units ="pt"),
      panel.spacing.x = unit(15,units = "pt"), 
      panel.spacing.y = unit(5,units = "pt"), 
      strip.placement = "inside",
      panel.border = element_rect(fill = NA),
      #panel.border = element_blank(),
      strip.text.x = element_text(face = "bold", hjust = 0.5, vjust = 1, margin = margin(t =5, b = 5, unit = "pt"), size = 10),
      strip.text.y.left = element_text(angle = 0),
      axis.text = element_text(size = 10, colour = "black"),
      axis.line.x.bottom = element_line(), 
      legend.position = "top",
    ) +
    scale_y_continuous(labels = scales::percent_format(accuracy = 1),n.breaks = 4, breaks = waiver(), limits = c(0,0.8) )

ggsave(plot = ref_comparison_boxplot, filename =  here::here("ALS_SC/plots/deconvolution_reference_compare_boxplot.pdf"), width = 9, height = 7)

ref_comparison_plot 
```

```{r}

# compare spinal cord segments for shared donors
sp_compare <- 
  all_res %>%
  filter(tissue != "Thoracic_Spinal_Cord") %>% #, ) %>%
  select(donor, tissue, cell, deconv, tool, ref, disease) %>%
  pivot_wider(names_from = tissue, values_from = deconv, values_fn = first )


sp_compare %>%
  drop_na() %>%
  filter(ref == "mathys", tool == "dtangle" ) %>%
  ggplot(aes(x = Cervical_Spinal_Cord, y = Lumbar_Spinal_Cord)) +
  geom_point(aes(colour = disease)) +
  facet_wrap(~cell, scales = "free") +
  geom_abline(slope = 1, intercept = 0) +
  theme_classic() +
  ggpubr::stat_cor()


```



Correlate estimates between cell-types for cases and controls

```{r}

cell_levels = c("oligodendrocytes",  "neurons",  "astrocytes", "pericytes" , "endothelial",     "microglia" )

make_cor_matrix <- function(data,  disease_, title_){
  tissue_ <- "Cervical_Spinal_Cord"
  cols <- colorRampPalette(brewer.pal(6,name="RdYlBu"))(72)
  brks <- seq(-1,1,length.out=72)  
  # fix ordering of cell types to cell levels
  cell_types <- as.character(na.omit(unique(data$cell) ))
  cell_types <- cell_levels[ cell_levels %in% cell_types] 
  
  data %>%
    #drop_na() %>%
    filter(tissue == tissue_, disease == disease_) %>%
    select("sample", "cell", "deconv", "disease", "tissue") %>%
    tidyr::pivot_wider(names_from=cell, values_from = deconv, values_fn = first ) %>%
    select("sample", all_of(cell_types) ) %>%
    column_to_rownames("sample") %>%
    as.matrix() %>%
    cor(method="pearson") %>%
  pheatmap(treeheight_row=0, treeheight_col=0, border_color="black", color=rev(cols), breaks=brks, cluster_rows=FALSE, cluster_cols = FALSE, legend=FALSE, main=title_,display_numbers = TRUE,show_colnames = FALSE, fontsize = 8,fontsize_number = 9, number_color = "black" ) %>%
    .[[4]]
}

make_cor_matrix(darmanis_music_res, disease_ = "Control", title_ = "Darmanis MuSiC - Control" )

grob_list <- 
list(
make_cor_matrix(darmanis_music_res, "Control", title_ = "Darmanis MuSiC - Control" ),
make_cor_matrix(darmanis_music_res, "ALS", title_ = "Darmanis MuSiC - ALS" ),
make_cor_matrix(darmanis_dtangle_res, "Control", "Darmanis dtangle - Control" ),
make_cor_matrix(darmanis_dtangle_res, "ALS", "Darmanis dtangle - ALS" ),
make_cor_matrix(mathys_music_df, "Control", title_ = "Mathys MuSiC - Control" ),
make_cor_matrix(mathys_music_df, "ALS", title_ = "Mathys MuSiC - ALS" ),
make_cor_matrix(mathys_dtangle_df, "Control",  "Mathys dtangle - Control" ),
make_cor_matrix(mathys_dtangle_df, "ALS", "Mathys dtangle - ALS" )
)

cor_plots <- patchwork::wrap_plots(grob_list, nrow = 4, heights = c(4,4,5,5)) 

ggsave(plot = cor_plots, filename =  here::here("ALS_SC/plots/deconvolution_celltype_cor.pdf"), width = 8, height = 8)

cor_plots

darmanis_music_res %>% filter(tissue == "Cervical_Spinal_Cord") %>% select(sample,disease) %>% distinct() %>% group_by(disease) %>% tally()


 # plot_layout(nrow = 2, heights = c(4,5))
```

